Update to D 2.067.1 FE

ibuclaw opened the pull request:
**WORK IN PROGRESS**

You may notice that there is a lacking of actual _merging_ of DMD here.  That's because it's not being committed yet.

There is a lot of catching up to do with regards to moving most things over to either free functions, double-dispatch visitors, or removing a whole range of stubs entirely!

This leaves some nice room in GDC to do a better clean-room redesign in some places.  Whilst starting from scratch is not really an immediate option, what I'll instead do it put up the scaffolding and slowly / surely move things across.

Some things I'd like to end up with are:
- [ ] Grouping functions into clearly named source files.
- [ ] Using `.cc` for sources that interact with the frontend, and...
- [ ] Using `.c` for sources for independent GDC code generation.
- [ ] Adding a `namespace gcc { namespace d { } }` for common codegen routines, or...
- [ ] Encapsulating codegen routines into the `Visitor` classes, and possibly in the process...
- [x] Remove `IRState` entirely, or merge it's routines as part of a bigger structure / `Visitor`, same goes also goes for...
- [ ] `dt_t*` being replaced with `tree` and remove the `dt_t**` functions, or merge `toDt` with the general `toElem` codegen routines.
- [ ] It would also be nice to remove `Symbol`, and `block`, but that will have to wait until 2.069 maybe.

Immediate things that need to be done:
- [x] Pass 2.067 the library unittests
- [x] Pass 2.067 the testsuite
- [ ] Pass 2.067 the library unittests with optimizations turned on.

ibuclaw commented:
Removed this fantastic hack:

```
#ifdef optimize
#undef optimize
#endif
```

Now use of the frontend `::optimize` method is no longer permitted in GDC.  Instead, use `ctfeInterpret` to get the value constant.

ibuclaw commented:
Removed `ArrayScope` helper, and instead of binding index/slices around `__dollar`, instead it's treated as a placeholder for the real length.  This has resulted in overall better codegen and back-end branch prediction.

ibuclaw commented:
Also removed all other codegen helpers that no longer have any benefit.

ibuclaw commented:
@jpf91 - Feel free to have a look here: https://github.com/ibuclaw/GDC/commit/f83b832f2bae192ec5c9d9a63a8bdeabff6e9bd8

I'll leave it at that for the moment, but thought I might get a second opinion on encapsulating routines that use `IRState` into `IRVisitor` before I go ahead and complete it.

jpf91 commented:
LGTM. I think that's the right approach in the long term.

ibuclaw commented:
OK - maintaining this PR is clearly not going to work in the long run, so I'm going to merge in some stuff early once there is a nice window open to do so.

ibuclaw committed:
Update root library to 2.067

ibuclaw committed:
Update frontend and druntime to 2.067

ibuclaw commented:
Started some work on the actual merge.  Currently keeping the method interface, rather than continuing on with visitor conversions, just for the sake of getting something using working first.

Changes made that I've knowingly done:
- Implemented `volatileLoad` and `volatileStore` based on previous PR.  I've verified that it honours volatile on x86_64.
- Removed `object.di` and renamed `object_.d` to `object.d`.  This actually happened in 2.068 development, but I ran into some really bad problems in 2.067.
- `cent`/`ucent` are supported explicitly in `d_type_for_mode`
- Fixed `build_struct_memcmp` to handle comparing types as arbitrary sized integers (eg: 256-bits).  This of course would only ever happen if the target supports the mode.
- `d_gcc_magic_module` has been relocated to `Target::loadModule`, and it now just pushes all modules into a list to scan after the first semantic pass rather than immediately after parsing.  This is because version conditions won't have been resolved yet and may wrongly be set to false (oops!).
- Fixed a useless dynamic cast at the beginning of a `catch { }` statement.  Thanks to @WalterBright for spotting it.

I think that's all for now.

ibuclaw commented:
Druntime unittests fails with a segfault.  Need to investigate why.

ibuclaw commented:
Tracked it down to https://github.com/D-Programming-Language/druntime/pull/941 which depended on compiler changes.  There should be a nicer way to keep track of things that cause silent corruption if not fixed in sync.

Other changes:
- Added `_d_delstruct` and `_aaGetY` runtime functions.
- Skip memcmp on zero sized structs for `is` and `==`.
- Avoid adding array bounds checks if it is CTFE-able verified to be within bounds.
- Set `hasDtor` flag in class typeinfo, if a dtor exists.

jpf91 commented:
> There should be a nicer way to keep track of things that cause silent corruption if not fixed in sync.

I don't think we can do automated testing for that (the autotester could try to compile/test druntime againts the latest released compiler as well, but as soon as we ave one breaking change all later pull request would fail testing).

We could probably ask all contributors to somehow mark their pull request in a special way (Requires: XYZ in the description or a comment). We could then maybe extend @MartinNowak dlang-bot to parse these comments and output a list for every druntime release?

ibuclaw commented:
> I don't think we can do automated testing for that (the autotester could try to compile/test druntime againts the latest released compiler as well, but as soon as we ave one breaking change all later pull request would fail testing).

Yeah.  There are a couple of things that I see as a recurring problem when updating.
- long gaps between releases make it difficult to track what changes depend on changes to codegen.
- heavy refactoring obscures actual changes in codegen.  git blame never tracks down the PR you're looking for.

ibuclaw committed:
Update phobos to 2.067

ibuclaw commented:
Internal changes to symbol emission:
- Remove `d_output_declaration_p`, move error checking into `FuncDeclaration::toObjFile`.
- Move duplicate symbol detection out of `d_finish_compilation` and into `d_comdat_linkage`.  Mark duplicates as `DECL_ABSTRACT_P`.
- Likewise mark templates as `DECL_ABSTRACT_P` when `!flag_emit_templates`.
- Don't send `DECL_ABSTRACT_P` functions to `cgraph_node::finalize_function` where it will be queued for emission (fixes broken phobos unittest build in std.parallelism).
- Move `isNonRoot` logic into `d_finish_function`.  Mark non-root functions as non-static and external.  Remove hacky `force_p` parameter from `FuncDeclaration::toObjFile`.

With the last item, cross-module inlining should allow be possible...

MartinNowak commented:
> We could then maybe extend @MartinNowak dlang-bot to parse these comments and output a list for every druntime release?

We're using the same regex across a few services, and it works fairly nice b/c people get something back when mentioning the issue.
[github's bugzilla integration](https://github.com/github/github-services/blob/2e886f407696261bd5adfc99b16d36d5e7b50241/lib/services/bugzilla.rb#L155)
[dlang-bot](https://github.com/MartinNowak/dlang-bot/blob/33fbf7c8ec96358c2f94a6e82922a2c3faf364f9/source/app.d#L67)
[changelog generator](https://github.com/D-Programming-Language/tools/blob/5b21861aab0063c0be2923b94e5be29e62adeb5e/changed.d#L76)
In general it's pretty hard though to get people to reliably tag certain commits.

> Yeah. There are a couple of things that I see as a recurring problem when updating.
> 
> long gaps between releases make it difficult to track what changes depend on changes to codegen.

Long gaps on gdc's or dmd's side?
The solution to the merge problem is obvious, continuously integrate upstream code.
This is needed anyhow to get gdc/ldc compat feedback before we release a dmd feature.

ibuclaw commented:
> In general it's pretty hard though to get people to reliably tag certain commits.

I don't think that changing the behaviour of a runtime function would have made it to the changelogs. ;-)

> Long gaps on gdc's or dmd's side?

It was easier to keep up when it was just Walter. And I managed to find the time to test that ever bugzilla issue fixed also worked in gdc.  ;-)

> The solution to the merge problem is obvious, continuously integrate upstream code.

A bit difficult while I've still got many old glue methods to remove (and still using C++ frontend).

ibuclaw commented:
- Sorted out codegen bug when using runtime `va_arg` (the new `doFormat` implementation now works).

Just a note, because `extern(C)` variadic  functions promote types passed over `<...>`, I'm not sure how best to handle the compiler `va_arg` intrinsic.  Should probably warn people about using it in an `extern(C)` function with `float`, `bool`, or any integral types smaller than `int` (or make it an error).
- Found assertion in `std.math` (see https://github.com/D-Programming-Language/phobos/pull/3886)

Druntime and Phobos unittests now pass.

ibuclaw committed:
Update D2 testsuite to 2.067

ibuclaw commented:
Slowly getting there...

Compiler:
- Rebased on master.
- Implemented support for Bugzilla 13586
- DMD's `-verror=XXX` can now be controlled by GCC's `-fmax-errors=XXX`.
- The list created by `Target::loadModule` is now drained twice.  First after `semantic` for the benefit of CTFE, and again after `semantic3` incase any more modules were loaded by template instantiations.  Not ideal, but at least it's enough to avoid getting undefined references to intrinsics such as `alloca`.

Testsuite:
- Fixed lots of bad testsuite flags, such as passing files over `REQUIRED_ARGS` instead of `EXTRA_SOURCES`, not yet upstreamed.
- However, have upstreamed some problematic tests / files for things I have no intention of supporting: https://github.com/D-Programming-Language/dmd/pull/5355
- Improved DMD to GDC switch conversion in DejaGNU testsuite, now handle more cases.

```
# of expected passes        20692
# of unexpected failures    25
# of unresolved testcases   35
```

ibuclaw commented:
Back-ported `object` module from 2.068, also updating `rt.critical_`, `rt.dmain2`, and `rt.monitor_` - for the latter using `gcc.gthreads` with the new interface.

This solves problems with conflicting symbols in `object.d` and other phobos modules.

```
# of expected passes        20775
# of unexpected failures    6
# of unresolved testcases   11
```

Three remaining issues. One failing runnable test (sdtor.d) and  two runnable tests that are not compiling (testdt.d and xtest46.d).

ibuclaw commented:
- Now use `_d_arrayassign_l` and `_d_arrayassign_r`.
- Discovered that since forever (887ebcc1456aa519a4038f6ac4399789284d5cb8) we've been passing `_d_arrayctor` and `_d_arrayassign` parameters the wrong way.  I have no idea how this could of not been detected for so long (I'm sending fix to master).

Now at:

```
# of expected passes        20787
# of unexpected failures    2
# of unresolved testcases   3
```

Where the remainder are non-trivial fixes.
- sdtor.d fails to compare array of structs under optimization (memcmp), may need to rethink how we initialize holes.  Maybe by using an artificial field hidden from debug codegen, however this might have an effect on ABI?  Investigation needed.
- testdt.d throws an error with no location deep in the Type::toDt routines (can't initialize a static void array).  These should be silently converted to byte arrays, but it gives the excuse to re-write this code and remove `toDt` and `dt_t` altogether (see OP).

ibuclaw commented:
The testsuite now passes, see #147 and #148 for sdtor-related fixes.

ibuclaw commented:
OK, let's get this in.  Just to allow anyone to iron out any bugs in master before trickling it down to other branches.

jpf91 commented:
Great! Have you already tested ARM builds?

ibuclaw commented:
Not yet, but I imagine I should pull in the 2.068 druntime C bindings.

MartinNowak commented:
Yay!
BTW, could anyone update http://gdcproject.org/downloads/LATEST please?

ibuclaw commented:
No downloads are available yet.

jpf91 commented:
@ibuclaw this has introduced 40 new unresolved tests on semaphore. Are these all caused by the missing C++ compiler on semaphore?
https://semaphoreci.com/d-programming-gdc/gdc/branches/master/builds/61

ibuclaw commented:
Hmm, I have not come across that before in my local testing.

ibuclaw commented:
Nope, it was #160 that introduced the regression.

ibuclaw commented:
It happens because `FuncAliasDeclaration::toSymbol` doesn't set it's own `csym`.  It was likely a mistake on my part to assume that it would be set anyway.

ibuclaw commented:
Ah, I was looking at the wrong output.

I've added the following to our build commands:

```
grep -q FAIL gcc/testsuite/gdc/gdc.sum && false ||:
grep UNRESOLVED gcc/testsuite/gdc/gdc.sum ||:
```

I hope this should do the right thing™

MartinNowak commented:
> BTW, could anyone update http://gdcproject.org/downloads/LATEST please?

There is already a 5.2 w/ the 2.066.1 FE, but Travis-CI is still on 4.9.3.

ibuclaw commented:
Maybe wait for travis to roll out the next LTS version of Ubuntu? I'm a bit weary of ABI changes in g++-5 will start breaking stuff.  In particular the C++ tests in the testsuite.

MartinNowak commented:
> Maybe wait for travis to roll out the next LTS version of Ubuntu?

They didn't change that ever and would get into many more troubles by doing it, so you'll wait a very long time. If the latest gdc version isn't compatible with older g++ versions, then people at least know they have to use an older version and can explicitly select gdc-4.9.3.
The main idea behind dmd, dmd-beta, dmd-nightly, gdc, and ldc is that people can automatically test their library for compatibility with the latest released version.

ibuclaw commented:
> The main idea behind dmd, dmd-beta, dmd-nightly, gdc, and ldc is that people can automatically test their library for compatibility with the latest released version.

Ah, I thought you meant for building just ddmd.
